arangod/MMFiles/MMFilesCollection.cpp

```diff
@@ -2445,232 +2445,180 @@ void MMFilesCollection::invokeOnAllElements(
-                << _logicalCollection->name() << "'";
-            return TRI_ERROR_DEADLOCK;
-          }
-          LOG_TOPIC(TRACE, arangodb::Logger::FIXME)
-              << "waiting for write-lock on collection '"
-              << _logicalCollection->name() << "'";
-        } else if (++iterations >= 5) {
-          // periodically check for deadlocks
-          TRI_ASSERT(wasBlocked);
-          iterations = 0;
-          if (_logicalCollection->vocbase()->_deadlockDetector.detectDeadlock(
-                  _logicalCollection, true) == TRI_ERROR_DEADLOCK) {
-            // deadlock
-            _logicalCollection->vocbase()->_deadlockDetector.unsetWriterBlocked(
-                _logicalCollection);
-            LOG_TOPIC(TRACE, arangodb::Logger::FIXME)
-                << "deadlock detected while trying to acquire "
-                   "write-lock on collection '"
-                << _logicalCollection->name() << "'";
-            return TRI_ERROR_DEADLOCK;
-          }
-        }
-      } catch (...) {
-        // clean up!
-        if (wasBlocked) {
-          _logicalCollection->vocbase()->_deadlockDetector.unsetWriterBlocked(
-              _logicalCollection);
-        }
-        // always exit
-        return TRI_ERROR_OUT_OF_MEMORY;
-      }
+  int res = TRI_ERROR_NO_ERROR;
+  if (useDeadlockDetector) {
+    res = lockWriteWithDeadlockDetection(timeout); 
+  } else {
+    bool isLocked = _dataLock.tryWriteLock(timeout);
+    if (!isLocked) {
+      res = TRI_ERROR_LOCK_TIMEOUT;
     }
+  }
 
-    double now = TRI_microtime();
+  if (res == TRI_ERROR_LOCK_TIMEOUT) {
+    LOG_TOPIC(TRACE, arangodb::Logger::FIXME)
+        << "timed out after " << timeout << " s waiting for write-lock on collection '"
+        << _logicalCollection->name() << "'";
+  } else if (res == TRI_ERROR_DEADLOCK) {
+    LOG_TOPIC(TRACE, arangodb::Logger::FIXME)
+        << "deadlock detected while trying to acquire write-lock on collection '"
+        << _logicalCollection->name() << "'";
+  }
+  return res;
+}
 
-    if (waitTime == 0) {  // initialize times
-      // set end time for lock waiting
-      if (timeout <= 0.0) {
-        timeout = defaultLockTimeout;
+int MMFilesCollection::lockWriteWithDeadlockDetection(double timeout) {
+  int res = TRI_ERROR_NO_ERROR;
+  bool wasBlocked = false;
+  auto onBlock = [this, &res, &wasBlocked]() -> bool {
+    TRI_ASSERT(res == TRI_ERROR_NO_ERROR);
+    try {
+      // insert writer
+      if (_logicalCollection->vocbase()->_deadlockDetector.setWriterBlocked(_logicalCollection) == TRI_ERROR_DEADLOCK) {
+        // deadlock
+        res = TRI_ERROR_DEADLOCK;
+        return false;
       }
-      startTime = now;
-      waitTime = 1;
+      wasBlocked = true;
+        
+      LOG_TOPIC(TRACE, arangodb::Logger::FIXME)
+          << "waiting for write-lock on collection '" << _logicalCollection->name() << "'";
+      return true;
+    } catch (...) {
+      res = TRI_ERROR_OUT_OF_MEMORY;
```

Comment:
To maintain compatibility with the old implementation in MMFilesCollection.cpp I replaced.
In theory there may be different reasons for the exception, why are not handled differently here.
But looking at the code in the `try` block, I think the only way it can actually throw an exception is if the system *is* running out of memory (copying the collection name can fail because of OOM), and the call to `deadlockDetector.setWriterBlocked(_logicalCollection)` can also fail with OOM. So it's probably right as it is now, but it is not future-proof.